sdk/python/kfp/components/types/type_utils.py

```diff
@@ -229,37 +212,62 @@ def verify_type_compatibility(
         InconsistentTypeException if types are incompatible and TYPE_CHECK==True.
     """
 
-    # Generic "Artifact" type is compatible with any specific artifact types.
-    if not is_parameter_type(
-            str(given_type)) and (str(given_type).lower() == 'artifact' or
-                                  str(expected_type).lower() == 'artifact'):
-        return True
-
     # Special handling for PipelineTaskFinalStatus, treat it as Dict type.
     if is_task_final_status_type(given_type):
         given_type = 'Dict'
 
-    # Normalize parameter type names.
-    if is_parameter_type(given_type):
-        given_type = get_parameter_type_name(given_type)
-    if is_parameter_type(expected_type):
-        expected_type = get_parameter_type_name(expected_type)
-
-    types_are_compatible = _check_types(given_type, expected_type)
+    types_are_compatible = False
+    is_parameter = is_parameter_type(str(given_type))
+
+    # handle parameters
+    if is_parameter:
+        # Normalize parameter type names.
+        if is_parameter_type(given_type):
+            given_type = get_parameter_type_name(given_type)
+        if is_parameter_type(expected_type):
+            expected_type = get_parameter_type_name(expected_type)
+
+        types_are_compatible = check_parameter_type_compatibility(
+            given_type, expected_type)
+
+    # handle artifacts
+    elif isinstance(given_type, str):
+        # v1 compat
+        given_suffix = given_type.lower().split('.')[-1]
+        expected_suffix = expected_type.lower().split('.')[-1]
+        types_are_compatible = (given_suffix
+                                == expected_suffix) or 'artifact' in {
+                                    given_suffix, expected_suffix
+                                }
+
+    elif isinstance(given_type, dict) or isinstance(expected_type, dict):
+        # v1 compat
+        given_type_name = list(given_type.keys())[0]
+        expected_type_name = list(expected_type.keys())[0]
+        types_are_compatible = given_type_name == expected_type_name and given_type[
+            given_type_name] == expected_type[expected_type_name]
 
+    else:
+        # v2
+        types_are_compatible = (given_type.schema_title
```

Comment:
If we were to introduce schema version properly, then the type compatibility checks needs to take into account version as well. I would assume the version would follow semantic versioning and not guarantee compatibility between major versions.

Reply:
Very good point. I will add this logic.

It makes sense to me that `system.Artifact` would still be compatible with all other artifacts, irrespective of either version. Does that sound right? Also, can you think of a case where we would ever want to bump the version of `system.Artifact`?

Reply:
That's a really good question. If `system.Artifact` is meant to be the generic one, which was the designed this way, It feels to me version is meaningless for `system.Artifact`. We probably need more thinking on this. But for now, what you described sounds good to me.